File: sdk/typescript/rooch-sdk/src/crypto/mnemonics.test.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { describe, it, expect } from 'vitest'
import { isValidBIP32Path, isValidHardenedPath } from './mnemonics.js'

describe('isValidHardenedPath', () => {
  it('should return true for valid path with typical indices', () => {
    const path = "m/44'/784'/0'/0'/0'"
    const result = isValidHardenedPath(path)
    expect(result).toBe(true)
  })

  it('should return true for valid path with higher indices', () => {
    const path = "m/44'/784'/123'/456'/789'"
    const result = isValidHardenedPath(path)
    expect(result).toBe(true)
  })

  it('should return false for path with missing "m" prefix', () => {
    const path = "44'/784'/0'/0'/0'"
    const result = isValidHardenedPath(path)
    expect(result).toBe(false)
  })

  it('should return false for path with missing apostrophes', () => {
    const path = 'm/44/784/0/0/0'
    const result = isValidHardenedPath(path)
    expect(result).toBe(false)
  })

  it('should return false for path with non-numeric indices', () => {
    const path = "m/44'/784'/a'/b'/c'"
    const result = isValidHardenedPath(path)
    expect(result).toBe(false)
  })
})

describe('isValidBIP32Path function', () => {
  it("should return false for valid BIP32 path m/54'/784'/0'/0/0", () => {
    const path = "m/54'/784'/0'/0/0"
    expect(isValidBIP32Path(path)).toBe(false)
  })

  it("should return true for invalid BIP32 path n/54'/784'/0'/0/0", () => {
    const path = "n/54'/784'/0'/0/0"
    expect(isValidBIP32Path(path)).toBe(true)
  })

  it("should return true for invalid BIP32 path m/53'/784'/0'/0/0", () => {
    const path = "m/53'/784'/0'/0/0"
    expect(isValidBIP32Path(path)).toBe(true)
  })

  it("should return true for invalid BIP32 path m/54'/785'/0'/0/0", () => {
    const path = "m/54'/785'/0'/0/0"
    expect(isValidBIP32Path(path)).toBe(true)
  })

  it("should return false for valid BIP32 path m/74'/784'/1'/1/1", () => {
    const path = "m/74'/784'/1'/1/1"
    expect(isValidBIP32Path(path)).toBe(false)
  })
})

File: sdk/golang/rooch-sdk/crypto/mnemonics_test.go
package crypto

import (
	"testing"
)

func TestIsValidHardenedPath(t *testing.T) {
	tests := []struct {
		name     string
		path     string
		expected bool
	}{
		{
			name:     "valid path with typical indices",
			path:     "m/44'/784'/0'/0'/0'",
			expected: true,
		},
		{
			name:     "valid path with higher indices",
			path:     "m/44'/784'/123'/456'/789'",
			expected: true,
		},
		{
			name:     "path with missing m prefix",
			path:     "44'/784'/0'/0'/0'",
			expected: false,
		},
		{
			name:     "path with missing apostrophes",
			path:     "m/44/784/0/0/0",
			expected: false,
		},
		{
			name:     "path with non-numeric indices",
			path:     "m/44'/784'/a'/b'/c'",
			expected: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := IsValidHardenedPath(tt.path)
			if result != tt.expected {
				t.Errorf("IsValidHardenedPath(%s) = %v, want %v", tt.path, result, tt.expected)
			}
		})
	}
}

func TestIsValidBIP32Path(t *testing.T) {
	tests := []struct {
		name     string
		path     string
		expected bool
	}{
		{
			name:     "valid BIP32 path case 1",
			path:     "m/54'/784'/0'/0/0",
			expected: false,
		},
		{
			name:     "invalid BIP32 path with n prefix",
			path:     "n/54'/784'/0'/0/0",
			expected: true,
		},
		{
			name:     "invalid BIP32 path with different first index",
			path:     "m/53'/784'/0'/0/0",
			expected: true,
		},
		{
			name:     "invalid BIP32 path with different second index",
			path:     "m/54'/785'/0'/0/0",
			expected: true,
		},
		{
			name:     "valid BIP32 path case 2",
			path:     "m/74'/784'/1'/1/1",
			expected: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := IsValidBIP32Path(tt.path)
			if result != tt.expected {
				t.Errorf("IsValidBIP32Path(%s) = %v, want %v", tt.path, result, tt.expected)
			}
		})
	}
} 